### ISSUE TITLE
(Implementation question) Multiheaded Attention

### ISSUE THREAD

**Issue by DaDucking** [RvuvuzelaM/self-attention-ppo-pytorch#5]
Hi @RvuvuzelaM , I just studied your code and realised that you have implemented multiheaded attention differently from the authors. Do you mind explaining the theory of your implementation of multiheaded attention? 

The standard implementation of scaled product attention uses softmax( Q transposed * K / sqrt(size)), however your implementation permutes q,k,v, into (0,2,3,1) which swaps channel with height and weight before going into a Q transposed * K and so on. Mathematically, the tensors end up being the same shape but do you mind explaining what is the meaning of each step in doing it this way? How does a pure matrix multiplication behaves as an attention layer and are there any learnable parameters in your attention layer? 

Thanks!

**Comment by MarcoMeter** [comment]
I'd be interested to know why queries, keys and values are the same.
